ExerciseController

iv-nav opened the pull request.

iv-nav committed:
test pull request

Commit:
SampleTaskController

Commit:
Merge branch 'dev' of https://github.com/softority/stp-testing-api into dev

iv-nav committed:
ExerciseController without a validation + soft delete for Exercise table + Enum Json Converter

iv-nav committed:
Merge branch 'dev' of https://github.com/softority/stp-testing-api into dev

iv-nav commented on Stp.TestingApi/EnumMapper.cs:
```diff
@@ -0,0 +1,17 @@
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Stp.TestingApi
```
I have planned to use mapper. But we need 
to discuss how it will work.

m-ratnikov commented on Stp.Data/Entities/Exercise.cs:
```diff
@@ -0,0 +1,23 @@
+    public class Exercise
+    {
+        public long Id { get; set; }
+        public string TestSectionId { get; set; }
```
why string? the type should be long

m-ratnikov commented on Stp.Data/Entities/Exercise.cs:
```diff
@@ -0,0 +1,23 @@
+        public long ExerciseCategoryId { get; set; }
+        public string Name { get; set; }
+        public string Description { get; set; }
+        public long Type { get; set; }
```
why long? Use enum

m-ratnikov commented on Stp.Data/Entities/Exercise.cs:
```diff
@@ -0,0 +1,23 @@
+        public long Type { get; set; }
+        public int Points { get; set; }
+        public int DurationMinutes { get; set; }
+        public int Complexity { get; set; }
```
why int? Use enum

m-ratnikov commented on Stp.Data/TestingDbContext.cs:
```diff
@@ -16,6 +16,36 @@ public TestingDbContext(DbContextOptions<TestingDbContext> options)
+                    case EntityState.Deleted:
+                        entry.State = EntityState.Modified;
+                        entry.CurrentValues["IsDeleted"] = true;
+                        break;
```
in terms of EF soft delete will mean updating the entity (setting IsDeleted = true). So, tracking states "Added" and "Deleted" is not sufficent.

m-ratnikov commented on Stp.TestingApi/Contracts/ExerciseDto.cs:
```diff
@@ -0,0 +1,33 @@
+        [EnumMember(Value = "Medium")]
+        Medium,
+        [EnumMember(Value = "High")]
+        High
```
use [Display(Name=...)] attribute instead

m-ratnikov commented on Stp.TestingApi/Controllers/ExerciseController.cs:
```diff
@@ -0,0 +1,185 @@
+        [HttpGet(nameof(GetExercisesByCategory))]
+        public IEnumerable<ExerciseDto> GetExercisesByCategory(long exerciseCategoryId)
+        {
+            var res = _db.ExerciseList
+                .ToList().Where(x => x.ExerciseCategoryId == exerciseCategoryId)
+                .Select(x => new ExerciseDto()
+                {
+                    Id = x.Id,
+                    Name = x.Name,
+                    Description = x.Description,
+                    Points = x.Points,
+                    DurationMinutes = x.DurationMinutes, 
+                    Complexity = (ExerciseComplexity)x.Complexity
+                });       
```
ToList function should be called the last.
Calling ToList before Select and Where clauses leads to generating SQL like "select * from Exercise" which is not effective.
EF converts Linq query to SQL expression before calling materialization function (like ToList, ToArray, FirstOrDefault etc)

m-ratnikov commented on Stp.TestingApi/Controllers/ExerciseController.cs:
```diff
@@ -0,0 +1,185 @@
+
+        /* https://localhost:5001/api/Exercise/AddExercise */
+        [HttpPost(nameof(AddExercise))]
+        public IActionResult AddExercise(long exerciseCategoryId, [FromBody]ExerciseDto exercise)
```
Function should return Exercise object, not IActionResult

m-ratnikov commented on Stp.TestingApi/Controllers/ExerciseController.cs:
```diff
@@ -0,0 +1,185 @@
+
+        private Exercise FindExercise(long id)
+        {
+            var res = _db.ExerciseList
```
It's OK, but when finding by PK, it's better to use Find function, which can be more effective then FirstOrDefault in some cases.

m-ratnikov commented on Stp.TestingApi/Controllers/ExerciseController.cs:
```diff
@@ -0,0 +1,185 @@
+            return res;
+        }
+
+        private Exercise FindExerciseIgnoringFilters(long id)
+        {
+            var res = _db.ExerciseList.IgnoreQueryFilters()
+                .Where(x => (x.Id == id))                
+                .FirstOrDefault();
+
+            return res;
+        }
+    }
```
Funtion is not used. For which reason it has been created?

m-ratnikov commented on Stp.TestingApi/EnumMapper.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace Stp.TestingApi
+{
+    public static class EnumMapper
+    {
+        public static int EnumToInt(ExerciseComplexity theEnum)
+        {
+            return 0; 
+        }
+
+    }
```
for what reason it created? ususal cast ((int)ExerciseComplexity.High) allows to get int value of the enum member.

m-ratnikov commented on Stp.TestingApi/Startup.cs:
```diff
@@ -35,6 +39,12 @@ public void ConfigureServices(IServiceCollection services)
+
+            services.AddMvc().AddJsonOptions(options =>
+                {
+                    options.JsonSerializerOptions.Converters.Add(new JsonStringEnumConverter());
```
is it really neccessary?

iv-nav committed:
+Exercise table
+Exercise controller without validation
+Soft delete
+MultichoiceExerciseAnswer table and the controller
+Eager loading of Answers

iv-nav committed:
comments removed

m-ratnikov commented on Stp.TestingApi/Controllers/ExerciseController.cs:
```diff
@@ -166,20 +179,8 @@ public IActionResult DeleteExercise(long exerciseId)
-                .FirstOrDefault();
-
-            return res;
+            return _db.ExerciseList.Find(id);
```
Function is seems redundant. It incapsulates only one simple operation

m-ratnikov commented on Stp.TestingApi/Controllers/MultichoiceExerciseAnswerController.cs:
```diff
@@ -0,0 +1,92 @@
+
+        /* https://localhost:5001/api/MultichoiceExerciseAnswer/AddExerciseAnswer */
+        [HttpPost(nameof(AddExerciseAnswer))]
+        public MultichoiceAnswerDto AddExerciseAnswer(long exerciseId, [FromBody]MultichoiceAnswerDto answerDto)
```
It is neccessary to check whether Exercise with exerciseId exists

m-ratnikov commented on Stp.TestingApi/Controllers/MultichoiceExerciseAnswerController.cs:
```diff
@@ -0,0 +1,92 @@
+
+            if (answer == null)
+            {
+                return BadRequest();
```
Here and in other similar places add error message like $"Answer with Id={answerId} not found"

m-ratnikov commented on Stp.TestingApi/Controllers/MultichoiceExerciseAnswerController.cs:
```diff
@@ -0,0 +1,92 @@
+            return Ok();
+        }
+
+        private MultichoiceExerciseAnswer FindAnswer(long id)
+        {
+            return _db.MultichoiceAnswerList.Find(id);
+        }
```
Function seems redundant

iv-nav committed:
errors fixed
